### DISCUSSION TITLE
batch insert is very slow

### DISCUSSION THREAD

**Discussion by laopeng2021** [duckdb/duckdb#3433] [General]
Hello.
I use duckdb to write 1000 pieces of data. Write speed is very slow. And there is a big difference in the length of each write.
Moreover, the batch mode of JDBC in duckdb is not supported.
Please ask if there is a good method for batch writing?

The code is as follows:
public static void insertToDuckdb() {
		try {
			Connection conn=null;
			 try {
				 Class.forName("org.duckdb.DuckDBDriver");
				 Properties ro_prop = new Properties();
					ro_prop.setProperty("duckdb.read_only", "False");
					conn = java.sql.DriverManager.getConnection("jdbc:duckdb:F://duckdb//test47.db",ro_prop);
			 }catch(Exception e) {
				 e.printStackTrace();
			 }
			 conn.createStatement()
		      .executeUpdate("create table testtable1 (TYPE_ID int, TYPE_DESC varchar(32), primary key(TYPE_ID))");

		    PreparedStatement localPreparedStatement1 = conn.prepareStatement("insert into testtable1 values(?, ?)");
		    int k;
		    long t1=System.currentTimeMillis();
		    for (int i = 1; i <= 1000; ++i) {
		    	 long t12=System.currentTimeMillis();
		      String str1 = "Type-" + i;
		      k = i;
		      localPreparedStatement1.setInt(1, k);
		      localPreparedStatement1.setString(2, str1);
		      localPreparedStatement1.execute();
		      long t13=System.currentTimeMillis();
		       System.out.println("Execution of line "+i+" took "+(t13-t12)+"milliseconds");
		      localPreparedStatement1.clearParameters();
		    }
		    localPreparedStatement1.close();
		    long t2=System.currentTimeMillis();
		    System.out.println("Total time："+(t2-t1));
		}catch(Exception e) {
			e.printStackTrace();
		}
		
	 
	}

The execution time is as follows：
.
.
.
Execution of line 154 took  37 milliseconds
Execution of line 155 took  513 milliseconds
Execution of line 156 took  512 milliseconds
Execution of line 157 took  503 milliseconds
Execution of line 158 took  291 milliseconds
Execution of line 159 took  60 milliseconds
Execution of line 160 took  61 milliseconds
Execution of line 161 took  184 milliseconds
Execution of line 162 took  58 milliseconds
Execution of line 163 took  80 milliseconds
Execution of line 164 took  65 milliseconds
.
.
.
Execution of line 988 took  16 milliseconds
Execution of line 989 took  15 milliseconds
Execution of line 990 took  282 milliseconds
Execution of line 991 took  501 milliseconds
Execution of line 992 took  264 milliseconds
Execution of line 993 took  15 milliseconds
Execution of line 994 took  15 milliseconds
Execution of line 995 took  16 milliseconds
Execution of line 996 took  15 milliseconds
Execution of line 997 took  15 milliseconds
Execution of line 998 took  16 milliseconds
Execution of line 999 took  15 milliseconds
Execution of line 1000 took  16 milliseconds
Total time：56386 milliseconds

**Comment by hannes** [comment]
Have a look at the appender api, looping a prepared statement is not a great way to insert large amounts of data into DuckDB

**Reply by laopeng2021** [reply]
Excuse me, is there a specific example for reference?

I don't know what to do. thank you.

**Comment by Alex-Monahan** [comment]
Here is our C client Appender documentation. Let us know if this helps!
https://duckdb.org/docs/api/c/appender

**Reply by laopeng2021** [reply]
thanks, appender is useful for me!
